### ISSUE TITLE
Enumerate all spanning trees

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1827]
**What is the feature or improvement you would like to see?**
An algorithm to generate all spanning trees. Since there are a very large number of spanning trees (see [Kirchhoff's them](https://en.wikipedia.org/wiki/Kirchhoff%27s_theorem)), it would make sense to implement something akin to Python's generators, i.e. a data structure that can be used for the _resumable_ generation of the trees.

**Use cases for the feature**

??

This has been [requested on StackOverflow](https://stackoverflow.com/q/69794497/695132). 

**References**

 - [Chakraborty et al., Algorithms for generating all possible spanning trees of a simple undirected connected graph: an extensive review](https://doi.org/10.1007/s40747-018-0079-7)
 - The Art of Computer Programming, Volume 4, Fascicle 4 contains an algorithm.

**Comment by iosonofabio** [comment]
Good point. This might actually be a good test case to implement a lazy iteration across languages and see what the pain points are...

**Comment by SMikhailE** [comment]
@szhorvat, thank you for the feature request. 

I am working with different problems in which spanning trees can be used. For example, seamless navigation indoor/outdoor building or a copula model for risk assessment.

Some of the bottlenecks in the algorithm are related to spanning trees. The minimum spanning tree is not suitable for all cases, and we need to look for the next candidate from the set of spanning trees. It would be very convenient to have an igraph function to generate a complete set of spanning trees. 

Unfortunatlly, my experiments with breadth and depth search (graph.bfs () and graph.dfs() functions respectively) does not provide a complete set of spanning tree yet.

**Comment by szhorvat** [comment]
@SMikhailE Give that even relatively small graphs have more spanning trees than what can be reasonably generated, would a function to sample them uniformly be useful to you?

**Comment by SMikhailE** [comment]
@szhorvat yes it is. Such function will be a useful tool, because, for example, small adjacency matrices (5x5) can be effectively used to copula modeling.
